tempesta_fw/http.c

```diff
@@ -2641,127 +2767,326 @@ tfw_http_hm_drop_resp(TfwHttpResp *resp)
 }
 
+static void
+tfw_http_skb_queue(TfwHttpMsg *hm, struct sk_buff *skb,
+		   tfw_http_parse_stage_t stage)
+{
+	if (!tfw_http_msg_is_streamed(hm)) {
+		ss_skb_queue_tail(&hm->msg.head_skb, skb);
+		return;
+	}
+
+	switch (stage)
+	{
+	case TFW_HTTP_PARSE_HEADERS:
+		ss_skb_queue_tail(&hm->msg.head_skb, skb);
+		break;
+	case TFW_HTTP_PARSE_BODY:
+		ss_skb_queue_tail(&hm->msg.body_skb, skb);
+		break;
+	case TFW_HTTP_PARSE_TRAILER:
+		ss_skb_queue_tail(&hm->msg.trailer_skb, skb);
+		break;
+	default:
+		BUG();
+	}
+}
+
+static int
+tfw_http_stream_req(TfwCliConn *cli_conn, TfwHttpReq *req)
+{
+	TfwSrvConn *conn = (TfwSrvConn *)cli_conn->conn_stream;
+	int r = 0;
+
+	if (!conn)
+		return 0;
+
+	TFW_DBG2("%s: resp=[%p]\n", __func__, req);
+
+	spin_lock(&conn->fwd_qlock);
+
+	if ((TfwHttpReq *)conn->msg_sent != req)
+		goto err;
+	if (!tfw_srv_conn_get_if_live(conn)) {
+		r = -EINVAL;
+		goto err;
+	}
+	tfw_cli_conn_mod_katimer((TfwCliConn *)req->conn);
+	r = __tfw_http_conn_stream((TfwConn *)conn, (TfwHttpMsg *)req);
+	tfw_srv_conn_put(conn);
+err:
+	spin_unlock(&conn->fwd_qlock);
+
+	return r;
+}
+
+/**
+ * Parse and add a new skb into current message.
+ *
+ * Before the call @skb is not attached to the message. If parsing succeed,
+ * @skb is either attached to the message and @skb pointer is set to NULL;
+ * or it's split into two parts, first is attached to the message, second is
+ * returned in @skb pointer.
+ */
+static int
+tfw_http_parse_skb(TfwConn *conn, struct sk_buff **skb, unsigned int *off,
+		   TfwFsmData *data_up)
+{
+	TfwHttpMsg *hm = (TfwHttpMsg *)conn->msg;
+	TfwHttpParser *parser = &hm->parser;
+	bool client = TFW_CONN_TYPE(conn) & Conn_Clnt;
+	ss_skb_actor_t actor = client ? tfw_http_parse_req
+				      : tfw_http_parse_resp;
+	size_t proxy_buff_sz = client ? tfw_cfg_cli_proxy_buff_sz
+				      : tfw_cfg_srv_proxy_buff_sz;
+	tfw_http_parse_stage_t stage;
+	unsigned int data_off;
+	int r = TFW_BLOCK;
+
+	/*
+	 * Process/parse data in the SKB.
+	 * @data_off points at the start of data for processing.
+	 * @off is the current offset of data to process in
+	 * the SKB. After processing @off points at the end
+	 * of latest data chunk. However processing may have
+	 * stopped in the middle of the chunk. Adjust it to point
+	 * to the right location within the chunk.
+	 */
+	stage = tfw_http_parse_stage(hm);
+	parser->skb = *skb;
+	data_off = *off;
+	r = ss_skb_process(*skb, off, actor, hm);
+	*off -= parser->to_go;
+	hm->msg.len += *off - data_off;
+	parser->skb = NULL;
+
+	if (client)
+		TFW_ADD_STAT_BH(*off - data_off, clnt.rx_bytes);
+	else
+		TFW_ADD_STAT_BH(*off - data_off, serv.rx_bytes);
+	TFW_DBG2("%s parsed: remain_len=%u parsed=%u msg_len=%zu"
+		 " ver=%d res=%d\n",
+		 (client ? "Request" : "Response"), (*skb)->len - *off,
+		 *off - data_off, hm->msg.len, hm->version, r);
+
+	if (unlikely(r == TFW_BLOCK))
+		return TFW_BLOCK;
+	if (hm->msg.len >= proxy_buff_sz)
+		hm->flags |= TFW_HTTP_F_STREAM;
```

Comment:
I still don't see how the comment https://github.com/tempesta-tech/tempesta/pull/1012#discussion_r187219535 is addressed. If a message is in streaming mode and we have buffer size configured as e.g. 10KB, then how do you guarantee that the socket receive queue won't exceed the size? Well, this is guaranteed by the fact that we immediately process each ingress skb, so a client can just flood us and we'll be forwarding unlimiteed amount of data.